### ISSUE TITLE
Report on per-thread states (Running, Sleeping, etc)

### ISSUE THREAD

**Issue by PeterZaitsev** [ncabatoff/process-exporter#41]
Hi,

Here is VMSTAT Data I see:

root@rocky:/var/lib/mysql# vmstat 10
procs -----------memory---------- ---swap-- -----io---- -system-- ------cpu-----
 r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
 7  1      0 60208724 249524 13975188    0    0   446  1264    1    0 10  2 84  5  0
 9  2      0 60183048 249524 14000796    0    0 45019 71400 28571 120740 11  2 83  4  0
19  0      0 60151924 249532 14028544    0    0 45933 58183 30071 124420 11  2 84  3  0
 5  0      0 60129124 249532 14054136    0    0 43394 59277 26325 119453 10  2 85  3  0
 8  0      0 60096532 249536 14081696    0    0 47411 73015 30156 124027 12  2 81  5  0
 7  0      0 60073072 249536 14108168    0    0 42472 59339 24350 117616 10  2 85  3  0

This shows there is certain number of processes always runable.  I would expect similar results from process_exporter, yet I see:

![image](https://user-images.githubusercontent.com/3651746/40272332-84a888cc-5b79-11e8-8088-a887c63728bd.png)

If I simplify this to prometheus expression:

max_over_time(namedprocess_namegroup_states{instance=~"rocky", state="Running"}[30s])

I still have value 0 returned for basically all processess:

![image](https://user-images.githubusercontent.com/3651746/40272354-deeb35f0-5b79-11e8-8a8d-df3a7a4f28a0.png)

This is:

root@rocky:/var/lib/mysql# uname -a
Linux rocky 4.4.0-124-generic #148-Ubuntu SMP Wed May 2 13:00:18 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

I wonder if this is some issue with kernel reporting or with information capture

**Comment by PeterZaitsev** [comment]
Hi,
Looks like same thing.  Note In this case I'm running MySQL Benchmark so I kind of expect MySQL to be out there in the top.   My theory could be main MySQL thread indeed sleeps most of the time and for this metric other thread statuses are not accounted for, while they are in VMSTAT

**Comment by ncabatoff** [comment]
If you provide instructions to reproduce I'll have a look.  Otherwise:
- try decreasing scrape interval (though you may need to make the config more specific to the mysql procs for process-exporter to keep up)
- try the -threads argument to process-exporter

**Comment by PeterZaitsev** [comment]
Looking more into it.  

Running "stress -c 8"  Which creates 8 separate processess spinning on CPU I see: 

namedprocess_namegroup_states{groupname="stress",state="Running"} 8

So it works properly for process 


In general it does not seems like States are counted per thread even if --threads argument is specified:

namedprocess_namegroup_states{groupname="mysqld",state="Other"} 0
namedprocess_namegroup_states{groupname="mysqld",state="Running"} 0
namedprocess_namegroup_states{groupname="mysqld",state="Sleeping"} 1
namedprocess_namegroup_states{groupname="mysqld",state="Waiting"} 0
namedprocess_namegroup_states{groupname="mysqld",state="Zombie"} 0

We can see only one process "sleeping"  here, which I guess is MySQL main thread, we have only one process

namedprocess_namegroup_num_procs{groupname="mysqld"} 1

But many threads:

namedprocess_namegroup_num_threads{groupname="mysqld"} 48

I think for practical purposes "sleeping", "running", "waiting" it is  better to count threads thank procs.

**Comment by ncabatoff** [comment]
I'm sorry, -threads was bad advice, I forgot that I never got around to adding the per-thread state metrics.  I will use this issue to track that work and rename it accordingly.

**Comment by ncabatoff** [comment]
Perhaps I've been thinking about this the wrong way.  The 'state' metrics have always been weak because they're gauges: mostly everything is sleeping and waiting, and threads often run only brief intervals, so catching anything in the running state is pretty rare.  #62 is probably a better way to provide visibility on how many threads are running or not, via the proposed cpu histograms.

If that takes care of measuring the running threads, we're left with just sleeping and waiting.  Sleeping is the default and can be inferred to be the state of what's not running or waiting.  For waiting, if I implement #18, we'll know not just how many threads are waiting, we'll see what they're waiting on.  In which case this bug possibly becomes a wontfix.

**Comment by PeterZaitsev** [comment]
Gauges are indeed very volatile but aggregated over period of time they can show good picture. 
There is a difference between CPU usage  and DiskIO and    Waiting.  Imagine I have 1 CPU core available.   The most I can see is the  process using 1 CPU core from CPU time standpoint.  However this can come from 1 running thread of from  10 threads  in the run queue.       It would be best if it would be easy to see the time process/thread spends in Run Queue but not being on CPU  or the time process spends waiting on disk IO but I do not think such stats are available and  "Running" and "Blocked" Gauges are at least somewhat reasonable estimates.

**Comment by ncabatoff** [comment]
Much as I hate to break backward compatibility, I'm going to do as you propose in this case, because I've done more testing and it looks like for threaded programs the per-state metrics aren't meaningful.  They only report the state of the first thread, which can be very deceptive.  The new load-generator cmd I just added always shows up as state S regardless of how many threads are in state D or R.

**Comment by ncabatoff** [comment]
v0.4.0 pre-release includes this change (state metrics count threads instead of procs.)
